### PR TITLE
feature(1354): Added lifetimes to Event + Tasks

### PR DESCRIPTION
**Pull request by ilikepi63**
Added in lifetimes for ppi Events + Tasks

### CODE REVIEW

**Comment by ilikepi63** [comment]
@Dirbaio I've implemented this. I realized that we potentially should be treating the tasks and events with their own lifetimes, but bounded to the driver's lifetime. 

This achieves what we want to achieve and removes the need to try and solve multiple references as the case was when we have a ppi_channel made up of a ppi_group disable_all task that is also references from the ppigroup. 

Hope this works!

**Comment by Dirbaio** [comment]
@ilikepi63: the extra lifetime param is not needed.

```rust
fn task_capture<'s: 'd>(&self) -> Task<'s>
```

behaves in practice the same as 

```rust
fn task_capture(&self) -> Task<'d>
```

`Task` is [covariant](https://doc.rust-lang.org/nomicon/subtyping.html), i.e. you can convert from `Task<'long>` to `Task<'short>`. So, you can call `.task_capture()` to get a `Task<'d>` and then implicitly convert it to a task with a shorter lifetime. You don't need an explicit `'s` in the signature to allow that.

**Comment by Dirbaio** [comment]
You're missing the lifetimes in PWM and GPIOTE. I think that should fix the CI issue.

File: embassy-nrf/src/ppi/mod.rs [hunk 1]
```diff
@@ -125,16 +126,16 @@ const REGISTER_DPPI_CONFIG_OFFSET: usize = 0x80 / core::mem::size_of::<u32>();
 /// When a task is subscribed to a PPI channel, it will run when the channel is triggered by
 /// a published event.
 #[derive(PartialEq, Eq, Clone, Copy)]
-pub struct Task(NonNull<u32>);
+pub struct Task<'d>(NonNull<u32>, PhantomData<&'d ()>);
 
-impl Task {
+impl<'d> Task<'_> {
```
Comment (Dirbaio) [inline embassy-nrf/src/ppi/mod.rs:131]:
```suggestion
impl<'d> Task<'d> {
```

File: embassy-nrf/src/ppi/mod.rs [hunk 2]
```diff
@@ -156,26 +160,29 @@ impl Task {
...
 
 /// Represents an event that a peripheral can publish.
 ///
 /// An event can be set to publish on a PPI channel when the event happens.
 #[derive(PartialEq, Eq, Clone, Copy)]
-pub struct Event(NonNull<u32>);
+pub struct Event<'d>(NonNull<u32>, PhantomData<&'d ()>);
 
-impl Event {
+impl<'d> Event<'_> {
```
Comment (Dirbaio) [inline embassy-nrf/src/ppi/mod.rs:171]:
```suggestion
impl<'d> Event<'d> {
```

File: embassy-nrf/src/ppi/ppi.rs [hunk 3]
```diff
@@ -3,12 +3,12 @@ use embassy_hal_common::into_ref;
 use super::{Channel, ConfigurableChannel, Event, Ppi, StaticChannel, Task};
 use crate::{pac, Peripheral};
 
-impl Task {
+impl<'d> Task<'_> {
```
Comment (Dirbaio) [inline embassy-nrf/src/ppi/ppi.rs:6]:
```suggestion
impl<'d> Task<'d> {
```

File: embassy-nrf/src/ppi/ppi.rs [hunk 4]
```diff
@@ -3,12 +3,12 @@ use embassy_hal_common::into_ref;
...
     fn reg_val(&self) -> u32 {
         self.0.as_ptr() as _
     }
 }
-impl Event {
+impl<'d> Event<'_> {
```
Comment (Dirbaio) [inline embassy-nrf/src/ppi/ppi.rs:11]:
```suggestion
impl<'d> Event<'d> {
```

**Comment by ilikepi63** [comment]
Thanks for the help on the lifetimes :) It seems I have quite a bit of reading to do on it. 

Looks like you were right on the missing stuff. Thanks!

**Review by Dirbaio — APPROVED** [review]
Looks great! Thanky ou

**Pull request merged**